types: add byte-comparable format support for native cql3 types

lkshminarayanan opened the pull request:
This PR introduces a new `comparable_bytes` class to add byte-comparable format support for all the [native cql3 data types](https://opensource.docs.scylladb.com/stable/cql/types.html#native-types) except `counter` type as that is not comparable. This implementation adheres to the byte-comparable format specification in https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/utils/bytecomparable/ByteComparable.md

Note that support for composite data types like lists, maps, and sets has not been implemented yet and will be made available in a separate PR.

Refs https://github.com/scylladb/scylladb/issues/19407

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16545/)
:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16545/)
#### Build Details:
- Duration: 38 min
- Builder: spider6.cloudius-systems.com

denesb commented on test/boost/comparable_bytes_test.cc:
```diff
@@ -204,3 +204,17 @@ BOOST_AUTO_TEST_CASE(test_float) {
+        uint64_t length = (uint64_t(1) << shift) - 1;
+        for (uint8_t sign_only_byte : {0, 0xFF}) {
+            bytes_ostream out;
+            encode_varint_length(length, sign_only_byte, out);
```
How is the test using this method, when it is only defined in the `.cc` file, I see no declaration in a header.

denesb commented on test/boost/comparable_bytes_test.cc:
```diff
@@ -16,6 +16,8 @@
 #include "utils/assert.hh"
```
Is this legacy comparison order also binding to us?
Does Cassandra or we use it in already existing sstable formats?

denesb commented on utils/comparable_bytes.cc:
```diff
@@ -397,6 +406,127 @@ struct to_comparable_bytes_visitor {
+        // The decimal 123.456 can be stored as unscaled value = 123456 and scale = 3.
+        // This needs to be normalized and converted to a mantissa that lies in [0.01, 1) and a base-100 exponent.
+        // such that the big_decimal value = mantissa [0.01, 1) * 100 ^ exponent.
+        // The decimal 123.456 becomes mantissa = 0.0123456 and base 100 exponent = 2 (i.e.) 0.0123456 * 100 ^ 2.
```
This could be very expensive to compute, to the point where it can cause stalls.
That said I don't know what we can do about it, I don't see how else we can make decimals byte comparable and Cassandra already made a decision, we have to follow if we want to remain compatible.
Maybe we should discourage (or outright forbid) using decimals in keys altogether.

This doesn't have to block this PR.

avikivity replied:
We can delay optimization of decimal until later. 

That said, it already caused huge trouble and can't be ignored. It will be easier to do it in a dedicated patch series.

avikivity commented on bytes_ostream.hh:
```diff
@@ -268,6 +268,43 @@ public:
+    // Writes the given integer value
+    template <std::integral int_type>
+    void write(const int_type& val) {
+        auto v = net::ntoh(val);
```
Actually need hton() since the input is in host byte order.

But I recommend write_be() instead into a byte array so you don't have to deal with the casting.

avikivity replied:
And also, to combine the biasing of signed types with the byteorder conversion, those aren't separate steps.

avikivity commented on bytes_ostream.hh:
```diff
@@ -268,6 +268,43 @@ public:
+    void write(managed_bytes_view v) {
+        if (v.empty()) {
+            return;
+        }
```
This isn't needed, the loop below is enough.

avikivity commented on utils/CMakeLists.txt:
```diff
@@ -17,6 +17,7 @@ target_sources(utils
     bloom_filter.cc
     buffer_input_stream.cc
     build_id.cc
+    comparable_bytes.cc
```
yes

avikivity commented on utils/comparable_bytes.cc:
```diff
@@ -15,12 +15,31 @@
+template <std::signed_integral T>
+static void convert_signed_fixed_length_integer(managed_bytes_view& src, bytes_ostream& out) {
+    constexpr auto sign_mask = T(1) << (sizeof(T) * 8 - 1);
+    out.write<T>(read_simple<T>(src) ^ sign_mask);
```
As mentioned before, should write_be() into a local buffer then copy it. The compiler will optimize it away to a single instruction.

avikivity commented on utils/comparable_bytes.cc:
```diff
@@ -15,12 +15,31 @@
+// To encode, invert the sign bit so that negative numbers are ordered before the positive ones.
+template <std::signed_integral T>
+static void convert_signed_fixed_length_integer(managed_bytes_view& src, bytes_ostream& out) {
+    constexpr auto sign_mask = T(1) << (sizeof(T) * 8 - 1);
```
I expect debug mode will complain about positive value converted to negative. Better use std::make_unsigned_t for all intermediates.

avikivity commented on utils/comparable_bytes.cc:
```diff
@@ -15,12 +15,31 @@
+
+    void operator()(const long_type_impl&) {
+        // Unimplemented
+        SCYLLA_ASSERT(false);
```
What's wrong with long?

avikivity replied:
Ah you mention it in the changelog.

avikivity commented on utils/comparable_bytes.cc:
```diff
@@ -23,6 +25,120 @@ static void convert_signed_fixed_length_integer(managed_bytes_view& src, bytes_o
+ *   2^20-1            encodes as           EFFFFF
+ *   2^20              encodes as           F0100000
+ *  -2^20              encodes as           100000
+ *   2^64-1            encodes as           FFFFFFFFFFFFFFFFFF
```
Did  you mean: 2^63-1?

mykaul commented:
Reminder that we've recently added the vector type:
```
commit 861fb58e14a5e679899657ef0fcb37b59310ecdf
Merge: c2b870ee54 a68bf6dcc1
Author: Avi Kivity <avi@scylladb.com>
Date:   Thu Feb 6 20:36:43 2025 +0200

    Merge 'vector: add support for vector type' from Dawid Pawlik
    
    This pull request is an implementation of vector data type similar to one used by Apache Cassandra.

```

avikivity commented on utils/comparable_bytes.cc:
```diff
@@ -23,6 +25,120 @@ static void convert_signed_fixed_length_integer(managed_bytes_view& src, bytes_o
+    // Discard the leading sign-only bytes
+    while (curr_byte == sign_only_byte && !src.empty()) {
+        curr_byte = read_simple<uint8_t>(src);
+    }
```
These loops are very wasteful. Use std::countl_one() and std::countl_zero() to figure out how many leading bytes can be discarded.

avikivity replied:
(and read the original long in one operation)

avikivity commented on utils/comparable_bytes.cc:
```diff
@@ -23,6 +25,120 @@ static void convert_signed_fixed_length_integer(managed_bytes_view& src, bytes_o
+
+    if (num_length_bits_to_write >= 8) {
+        // Write one byte full of length bits (i.e. inverted sign bits)
+        out.write<uint8_t>(~sign_only_byte);
```
Use a temporary byte array (of size 9) to avoid engaging the bytes_ostream machinery for every byte. It is heavy since it needs to check for out-of-space each time.

avikivity commented on utils/comparable_bytes.cc:
```diff
@@ -23,6 +25,120 @@ static void convert_signed_fixed_length_integer(managed_bytes_view& src, bytes_o
+    uint8_t bytes_with_sign_bit = sizeof(int64_t) - (1 + bytes_to_read);
+    const uint8_t sign_only_byte = int8_t(~length_bit) >> 7;
+    while (bytes_with_sign_bit--) {
+        out.write<uint8_t>(sign_only_byte);
```
Again, collect the results into a scalar and serialize the entire scalar with one operation.

lkshminarayanan replied:
Done.

denesb commented:
> Reminder that we've recently added the vector type:
> 
> ```
> commit 861fb58e14a5e679899657ef0fcb37b59310ecdf
> Merge: c2b870ee54 a68bf6dcc1
> Author: Avi Kivity <avi@scylladb.com>
> Date:   Thu Feb 6 20:36:43 2025 +0200
> 
>     Merge 'vector: add support for vector type' from Dawid Pawlik
>     
>     This pull request is an implementation of vector data type similar to one used by Apache Cassandra.
> ```

Do we allow this type to be part of keys?

avikivity commented on utils/comparable_bytes.cc:
```diff
@@ -23,6 +25,120 @@ static void convert_signed_fixed_length_integer(managed_bytes_view& src, bytes_o
+    uint8_t bytes_to_read = length_bits_in_curr_byte - 1; // -1 to account for the byte already read
+    if (bytes_to_read == 7) {
+        // check also the first bit in next byte for length bit
+        curr_byte = read_simple<uint8_t>(src);
```
This byte-by-byte reading is wasteful too. We could have something like this:

```c++
if (src.current_fragment.size() >= 9) {
   // decode directly from current fragment, then consume the number of bytes read from the view
} else {
   // decode byte by byte
}
```

Since I expect you'll need it elsewhere, you can encapsulate it into a helper function:

```c++
template <typename Func>
auto decode_opportunistically(
        managed_bytes_view& src,
        bytes_ostream& out,
        size_t wanted_fragment_size,
        Func func);
```

Where Func accepts a read_byte parameter that either does *p++ (when consuming from a single fragment) or calls read_simple<uint8_t> (when not).

avikivity commented on utils/comparable_bytes.cc:
```diff
@@ -131,6 +131,31 @@ static void decode_signed_long_type(managed_bytes_view& src, bytes_ostream& out)
+
+    // Peek the first byte to deduce sign.
+    // Positive values have their sign bit flipped in their encoded format.
+    uint8_t curr_byte = uint8_t(src[0]);
```
This would benefit from optimization when the current fragment is long enough.

avikivity commented on utils/comparable_bytes.cc:
```diff
@@ -131,6 +131,31 @@ static void decode_signed_long_type(managed_bytes_view& src, bytes_ostream& out)
+        out.write<uint_t>(read_simple<uint_t>(src) ^ sign_mask);
+    } else {
+        // invert all bytes;
+        out.write<uint_t>(~read_simple<uint_t>(src));
```
Better to read the input and write the output outside the if statement to reduce code footprint.

avikivity commented on test/boost/comparable_bytes_test.cc:
```diff
@@ -174,3 +174,33 @@ BOOST_AUTO_TEST_CASE(test_timestamp) {
+}
+
+BOOST_AUTO_TEST_CASE(test_double) {
+    byte_comparable_test(floating_point_test_data_generator<double>());
```
How do we test that the results are compatible with Cassandra?

lkshminarayanan replied:
There are currently no tests to verify that. We can probably add them when we implement the trie indexes that uses byte comparable format.

lkshminarayanan replied:
This is the only thing remaining now. 

There does not seem to be simple way to do this. I was trying to hack the Cassandra code to write out comparable bytes as values of a table coulmn but the latest version which has the support for byte comparable types writes sstable in nb format, which we do not support yet. I am trying another approach now where I'm generating values and directly writing into ScyllaDB to generate the sstables. I'll update as soon as I'm done.

avikivity commented on utils/comparable_bytes.cc:
```diff
@@ -131,6 +131,53 @@ static void decode_signed_long_type(managed_bytes_view& src, bytes_ostream& out)
+    while (pos > 0) {
+        pos -= 8;
+        out.write<uint8_t>(uint8_t((length >> pos) & 0xFF) ^ sign_only_byte);
+    }
```
In fact you have the size, just use a temporary array and write the entire array at once.

avikivity commented on utils/comparable_bytes.cc:
```diff
@@ -307,6 +364,32 @@ struct from_comparable_bytes_visitor {
 
+    // Decode a byte-comparable representation to the varint number.
+    void operator()(const varint_type_impl&) {
+        // Check the first byte to determine if the bytes were encoded using long or full-form encoding
```
Please push to a named function, for readability.

lkshminarayanan replied:
Done.

avikivity commented on utils/comparable_bytes.cc:
```diff
@@ -314,6 +333,38 @@ struct to_comparable_bytes_visitor {
+    // For time-based UUIDs (version 1), additional reordering is required
+    // to maintain time-based ordering in byte-comparable form.
+    void operator()(const uuid_type_impl&) {
+        uint64_t msb = read_simple<uint64_t>(serialized_bytes_view);
```
Again use named functions.

avikivity replied:
Just have operator() call a named function. Avoids having a huge class and easier to understand stack traces.

lkshminarayanan replied:
Done.

avikivity commented on utils/comparable_bytes.cc:
```diff
@@ -314,6 +333,38 @@ struct to_comparable_bytes_visitor {
+            out.write<uint64_t>(timeuuid_msb_to_comparable_bytes(msb));
+        } else {
+            // For non-time UUIDs, write the msb after shifting the version bits to the beginning
+            out.write<uint64_t>((version << 60) | ((msb >> 4) & 0x0FFFFFFFFFFFF000L) | (msb & 0xFFFL));
```
Better move out.write() outside the if statement.

avikivity commented on utils/comparable_bytes.cc:
```diff
@@ -314,6 +333,38 @@ struct to_comparable_bytes_visitor {
+        }
+
+        // Write the lsb
+        out.write(serialized_bytes_view, sizeof(uint64_t));
```
You're writing a constant. How did this pass the test?

avikivity commented on utils/comparable_bytes.cc:
```diff
@@ -406,6 +457,26 @@ struct from_comparable_bytes_visitor {
+        }
+
+        // Write the lsb
+        out.write(comparable_bytes_view, sizeof(uint64_t));
```
!?!?!?!

avikivity replied:
?

avikivity commented on utils/comparable_bytes.cc:
```diff
@@ -230,6 +230,38 @@ static uint64_t timeuuid_msb_from_comparable_bytes(uint64_t byte_comparable_msb)
+    // and all limbs except the most significant limb will have bits_per_limb.
+    // So, total bits = (num of limbs - 1) * bits per limb + bits in ms limb.
+    const auto& backend = value.backend();
+    const auto limb_count = backend.size();
```
Do we really need to access the internals?

avikivity replied:
We can consider switching to boost::multiprecision::cpp_dec_float types if it makes it easier.

avikivity commented on utils/comparable_bytes.cc:
```diff
@@ -527,11 +535,53 @@ struct to_comparable_bytes_visitor {
+        bool escaped = false;
+        while (!serialized_bytes_view.empty()) {
+            uint8_t curr_byte = read_simple<uint8_t>(serialized_bytes_view);
+            if (escaped) {
```
text is extremely common and also extremely unlikely to have zeros. So best have a fast-path that checks no zeros are there and just copy stuff, and a slow-path that operates byte-by-byte.

Also, loop on every fragment instead of every byte.

lkshminarayanan commented on test/boost/comparable_bytes_test.cc:
```diff
@@ -7,10 +7,113 @@
+
+    // test encode/decode
+    const auto test_data = gen.test_data();
+    const auto test_data_type = gen.data_type();
```
Ack. Done.

lkshminarayanan commented:
Changelog V2 :
 - Updated most methods to read and write bytes as a whole rather than doing it byte by byte. Use an intermediate buffer if a single read/write is not possible.
 - Removed unnecessary clutter from the testcases and directly pass the generated test data to the test method.

avikivity commented on bytes_ostream.hh:
```diff
@@ -268,6 +268,35 @@ public:
         write(bytes_view(reinterpret_cast<const signed char*>(ptr), size));
     }
 
+    // Writes the fragmented view
```
> bytes_ostream: overload write() to support more input types

Can narrow down to FragmentedView, now you're not supporting various integer types.

avikivity commented on test/boost/bytes_ostream_test.cc:
```diff
@@ -334,3 +334,34 @@ BOOST_AUTO_TEST_CASE(test_conversion_to_managed_bytes) {
+BOOST_AUTO_TEST_CASE(test_write_managed_bytes_view) {
+    // Generate data for test
+    int count = 25000;
+    managed_bytes source_bytes(managed_bytes::initialized_later(), sizeof(long) * count);
```
Ah, you write some stuff to the input later.

avikivity commented on types/comparable_bytes.cc:
```diff
@@ -8,19 +8,51 @@
+    // the sign mask to big-endian and XOR it directly with the serialized bytes to flip the sign bit in place.
+    static const auto sign_mask_be = seastar::cpu_to_be<unsigned_type>(unsigned_type(1) << (sizeof(unsigned_type) * 8 - 1));
+    unsigned_type flipped_value = read_simple_native<unsigned_type>(src) ^ sign_mask_be;
+    write_native_int<unsigned_type>(out, flipped_value);
```
Okay, I see you're operating 100% in big-endian.

Yet another way to do it is to avoid integers completely and read a byte array of the right size, flip the sign bit of the first byte, then write the byte array back. But the code is okay as is.

avikivity commented on types/comparable_bytes.cc:
```diff
@@ -36,21 +47,124 @@ static void convert_signed_fixed_length_integer(managed_bytes_view& src, bytes_o
+    std::array<char, 9> encoded_buffer;
+    if (num_bytes == 9) {
+        // An extra byte full of inverted sign bits has to be included in the comparable format
+        encoded_buffer[0] = negative_mask < 0 ? 0 : 0xFF;
```
This is super minor but I can't resist: encoded_buffer[0] = uint8_t(negative_mask) ^ 0xff

avikivity commented on types/comparable_bytes.cc:
```diff
@@ -592,11 +600,88 @@ struct to_comparable_bytes_visitor {
+    // However, the byte following the prefix is guaranteed to be FE or FF, making the encoding weakly prefix-free.
+    // Additionally, any such prefix sequence will compare as smaller than the value it prefixes,
+    // because any permitted separator byte will be smaller than the byte following the prefix.
     void operator()(const abstract_type& type) {
```
It's not good that we have a catch-all match here. Adding a new type will not cause a compile error.

Better to list the supported types one-by-one and forward them all to a function that handles them all in the same way.

avikivity commented on types/comparable_bytes.cc:
```diff
@@ -592,11 +600,88 @@ struct to_comparable_bytes_visitor {
-        on_internal_error(cblogger, fmt::format("byte comparable format not supported for type {}", type.name()));
+        if (!type.is_byte_order_equal()) {
+            on_internal_error(cblogger, fmt::format("byte comparable format not supported for type {}", type.name()));
+        }
```
And then this isn't needed.

avikivity commented on types/comparable_bytes.cc:
```diff
@@ -592,11 +600,88 @@ struct to_comparable_bytes_visitor {
+                // There are zeros that need to be escaped;
+                // Use a buffer to write the encoded value. The encoded bytes will have one extra byte for every
+                // zero sequence in the source. So, worstcase, it can only be 1.5 times the size of the source.
+                std::vector<signed char> buffer(size_t(std::ceil(double(curr_frag.size()) * 1.5)));
```
We might avoid the allocation by limiting the fragment processing size to a known value.

But at least use utils::small_vector (with an internal buffer of 1k) to avoid tiny allocations.

avikivity commented:
Looks good, but I'm concerned about test coverage. It's not just about the values, but also about how they are fragmented in the source.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16793/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16793/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16793/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/bytes_ostream_test
:small_blue_diamond: boost/comparable_bytes_test
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/7957/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16010/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4787/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16793/)
#### Build Details:
- Duration: 7 hr 54 min
- Builder: spider6.cloudius-systems.com

lkshminarayanan commented:
Changelog V3 :
- Byte encoder/decoder processes the data in fixed size chunks rather than whole fragments to avoid tiny allocations.
- Extract all logic into separate functions to keep the visitor classes clean

lkshminarayanan commented:
> Looks good, but I'm concerned about test coverage. It's not just about the values, but also about how they are fragmented in the source.

The test cases for natively ordered types - ascii, text, blob - include large test data that are fragmented. These are the only cases where we directly access fragments during encoding/decoding. For all other types we check and read the required number of types from source, so it should not cause any problems w.r.to. fragments.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17040/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17040/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17040/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/bytes_ostream_test
:small_blue_diamond: boost/comparable_bytes_test
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8162/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/4990/)
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16230/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17040/)
#### Build Details:
- Duration: 7 hr 52 min
- Builder: spider9.cloudius-systems.com

avikivity commented:
What about making sure the encoding is compatible with Cassandra?

lkshminarayanan commented:
> What about making sure the encoding is compatible with Cassandra?

Yes I'm working on that. I left a comment [here](https://github.com/scylladb/scylladb/pull/23541#discussion_r2066147324) :

This is the only thing remaining now. 

There does not seem to be simple way to do this. I was trying to hack the Cassandra code to write out comparable bytes as values of a table coulmn but the latest version which has the support for byte comparable types writes sstable in nb format, which we do not support yet. I am trying another approach now where I'm generating values and directly writing into ScyllaDB to generate the sstables. I'll update as soon as I'm done.

avikivity commented:
> > What about making sure the encoding is compatible with Cassandra?
> 
> Yes I'm working on that. I left a comment [here](https://github.com/scylladb/scylladb/pull/23541#discussion_r2066147324) :
> 
> This is the only thing remaining now.
> 
> There does not seem to be simple way to do this. I was trying to hack the Cassandra code to write out comparable bytes as values of a table coulmn but the latest version which has the support for byte comparable types writes sstable in nb format, which we do not support yet. I am trying another approach now where I'm generating values and directly writing into ScyllaDB to generate the sstables. I'll update as soon as I'm done.

Suggest to write a separate main entry point that calls those functions directly, and prints them to stdout.

michoecho reviewed (commented):
I just read the tests (and skimmed the implementation). I left some comments. The gist:
- Do we trust in the correctness? How do we know that the tests are good enough? Since the problem at hand is so self-contained, well-defined, and multi-cased, this might be a good place to make sure that the tests are *extra* good. Especially since we (in the entire project) are bad at systematic testing and it's a good practice opportunity.
- There's an arithmetic bug in one of the tests.
- Please don't introduce new random seeds independent from the main test seed.
- I'm surprised you bothered implementing decoders from `comparable_bytes` back into "normal" types, because there's no need for this transformation. (Other than maybe for debugging and for cross-checking in tests).

lkshminarayanan commented:
Changelog V4:
 - added tests to verify compatibility with cassandra
 - fixed an issue in `decode_signed_long_type()` that didn't trim redundant sign-byte for some values.

I somehow missed Michal's review. I’ll submit a new version to address his comments.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17278/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17278/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17278/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/bytes_ostream_test
:small_blue_diamond: boost/comparable_bytes_test
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16468/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5217/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8385/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17278/)
#### Build Details:
- Duration: 7 hr 6 min
- Builder: spider8.cloudius-systems.com

lkshminarayanan commented on test/boost/comparable_bytes_test.cc:
```diff
@@ -0,0 +1,493 @@
+
+    // generate some random test data
+    std::random_device rd;
+    const auto seed = rd();
```
Okay. I replaced all such usages with functions from tests::random.

bhalevy commented:
@Lakshmipathi @denesb anything holding up this PR, or is it ready for merge?

lkshminarayanan commented:
> @Lakshmipathi @denesb anything holding up this PR, or is it ready for merge?

@bhalevy Michal has suggested some improvements to the test case and I'm working on them.

lkshminarayanan commented on test/boost/comparable_bytes_test.cc:
```diff
@@ -0,0 +1,493 @@
+        }
+    }
+
+    // Generates test values by shifting bit(1) through all possible positions and then deriving
+    // multiple test cases from each value. This helps test edge cases and boundary conditions
+    // by covering values with different bit patterns across the entire range of the type.
+    auto num = int_type(1);
+    auto num_bits = sizeof(int_type) * 8;
+    std::vector<data_value> test_data;
+    test_data.reserve(num_bits * 4);
+    while (num_bits-- > 0) {
+        for (int_type n : std::initializer_list<int_type>{num, num - 1, ~num, ~(num - 1)}) {
+            if (!filter_func || filter_func(n)) {
+                test_data.push_back(create_data_value_func(n));
+            }
+        }
+        num <<= 1;
+    }
```
Ack.

lkshminarayanan commented on test/boost/comparable_bytes_test.cc:
```diff
@@ -0,0 +1,493 @@
+    test_data.reserve(test_data.size() + (20 * 4 * 4));
+    auto multiprecision_one = utils::multiprecision_int(1);
+    for (int shift = 1; shift <= 20; shift++) {
+        for (auto shift_prod : {64, 100, 256, 512}) {
```
Yes prod was not correct. I have renamed it to `shift_multiplier`.

lkshminarayanan commented on test/boost/comparable_bytes_test.cc:
```diff
@@ -0,0 +1,493 @@
+    // test uuids with random timestamp and clock sequences
+    for (auto i = 0; i < 20; i++) {
+        test_data.push_back(create_data_value(utils::UUID_gen::get_random_time_UUID_from_micros(std::chrono::milliseconds{dist(gen)})));
+    }
```
Ack. I have updated the test to use tests::random and also added the flipped bit comparison.

lkshminarayanan commented:
Changelog V5 : 
- use functions from `tests::random` instead of environment dependant random functions
- fixed a [bug](https://github.com/scylladb/scylladb/pull/23541#discussion_r2068694710) in `test_varint_length_encoding` test
- added more random testcases to uuid/timeuuid tests.

michoecho commented on test/boost/comparable_bytes_test.cc:
```diff
@@ -0,0 +1,493 @@
+    BOOST_REQUIRE_MESSAGE(std::ranges::is_sorted(test_items, [&test_data_type] (const test_item& a, const test_item& b) {
+        return test_data_type->compare(a.serialized_bytes, b.serialized_bytes) == std::strong_ordering::less;
+    }), "sorting items based on comparable bytes failed");
+}
```
You're right, sorting is enough here. Looking at my comment again, it's nonsense. I don't know what I was thinking.

Disregard this thread.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17504/)
:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17504/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17504/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/bytes_ostream_test
:small_blue_diamond: boost/comparable_bytes_test
:white_check_mark: - [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/16635/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5384/)
:white_check_mark: - [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8551/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17504/)
#### Build Details:
- Duration: 7 hr 4 min
- Builder: spider2.cloudius-systems.com

michoecho commented on bytes_ostream.hh:
```diff
@@ -268,6 +268,35 @@ public:
+        }
+
+        // copy rest of the fragments one by one into newly alloc'ed chunks
+        while (!v.empty()) {
+            auto this_size = std::min(v.current_fragment().size(), size_t(max_chunk_size()));
+            std::copy_n(v.current_fragment().begin(), this_size, alloc_new(this_size));
+            v.remove_prefix(this_size);
+        }
```
FragmentedView doesn't guarantee that `current_fragment()` is nonempty.
If `current_fragment()` is empty, this will become an infinite loop.

But why did you even add a new write primitive for this, instead of using the existing ones? I suggest
```c++
    template<FragmentedView View>
    void write(const View& v) {
        for (bytes_view f : fragment_range(v)) {
            write(f);
        }
    }
```

Adding new "low-level" methods has to be justified by some performance reasons, otherwise it adds space for bugs, like this one.

michoecho commented on bytes_ostream.hh:
```diff
@@ -268,6 +268,35 @@ public:
+
+    // Writes n bytes from the given fragmented view
+    template<FragmentedView View>
+    void write(View& v, size_t n) {
```
This method and the one above mutate the View (i.e. remove a prefix from it).
This is evil because that's a different behavior than for `bytes_view`.
Calling `managed_bytes_view x; out.write(x)` will modify `x`.
Calling `bytes_view x; out.write(x)` won't modify `x`.
That's begging for trouble.

Also, in general, a call like `out.write(x)` doesn't usually modify `x`. So I think this interface is surprising, in a bad way.

Either rename this method to `write_and_remove_prefix` (and the one above to `write_and_make_empty` or something), or modify those methods (and their callers) so that they take `View`/`const View&`.

lkshminarayanan replied:
I have removed the `write(View& v, size_t n)` entirely and changed `write(View& v)` to `write(View v)`. The callers now send a prefix to this method when required.

michoecho reviewed (changes requested):
I pointed out a bug added in `bytes_ostream`, please fix.

But aside from that, I see no reason to delay the PR.

Note: I haven't done an in-depth review of `comparable_bytes.cc`, because:
1. Avi and Botond did that before I had a chance.
2. From my point of view, this code is in flux anyway until trie indexes are merged, and I might have to make changes anyway to better integrate them, so I plan on reading it all again anyway after trie indexes are finished.

But the tests and the surface look okay to me.

Another note: I'm undecided on if we want to convert the entire key to comparable bytes at once.
It can be surprisingly expensive. If you have a big (tens of kilobytes) clustering key of type `blob`, with many zero bytes, just escaping the zero's can be the dominating cost of the query. Instead of having `comparable_bytes` be a wrapper around `managed_bytes`, we might want it to be a generator of fragments. But if we go into that direction, we'll just modify the code, no reason to think about it too hard for this PR.

avikivity commented:
> 
> Another note: I'm undecided on if we want to convert the entire key to comparable bytes at once. It can be surprisingly expensive. If you have a big (tens of kilobytes) clustering key of type `blob`, with many zero bytes, just escaping the zero's can be the dominating cost of the query. Instead of having `comparable_bytes` be a wrapper around `managed_bytes`, we might want it to be a generator of fragments. But if we go into that direction, we'll just modify the code, no reason to think about it too hard for this PR.

I believe this case is on one hand rare, and on the other hand can be mitigated with fancy SIMD (VPCOMPRESS/VPEXPAND). Incremental generation has better worst-case behavior but worse averge-case behavior. While we often sacrifice the average case to improve the worst case, I don't think it's needed here.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

#### Build Details:
- Duration: 2 min 52 sec
- Builder: spider6.cloudius-systems.com

yaronkaikov commented:
@lkshminarayanan @scylladb/scylla-maint I think this patch is breaking next -[Gating](https://jenkins.scylladb.com/job/scylla-master/job/next/9310/)

yaronkaikov commented:
> > @lkshminarayanan @scylladb/scylla-maint I think this patch is breaking next -[Gating](https://jenkins.scylladb.com/job/scylla-master/job/next/9310/)
> 
> How did it pass CI then?

it didn't even run :-(

nyh commented:
> #### 🟢 CI State: SUCCESS
> #### Build Details:
> 
>     * Duration: 2 min 52 sec
> 
>     * Builder: spider6.cloudius-systems.com

This run was supiciously fast :-) Obviously it didn't really run the tests. How come?

denesb commented:
Queueing again.

avikivity commented:
> Queueing again.

Last time the new tests failed in debug mode.

yaronkaikov commented:
@avikivity @denesb why did we queued this again, it's still failing Gating -[next](https://jenkins.scylladb.com/job/scylla-master/job/next/9315/)

lkshminarayanan commented:
> > ```
> > * Duration: 1 hr 48 min
> > ```
> 
> Was this the entire CI run? Isn't it too quick?
> 
> Also, why does the summary in the bottom (not in a comment) say that 4 test results are pending?

I don't see any tests run in the Jenkins log but somehow the CI posted success.

lkshminarayanan commented:
I found the issue. I have been exclusively using CMake and forgot to include the new files to `configure.py`. I have update the PR now. Hopefully the CI should succeed now.

lkshminarayanan commented:
The new CI build threw an exception : "01:01:40  [WARNING]: Failed to generate allure report: java.nio.file.FileAlreadyExistsException: /var/lib/jenkins/jobs/scylla-master/jobs/scylla-ci/builds/17850/archive/allure-report.zip
"

```
01:02:32  Also:   hudson.remoting.ProxyException: org.jenkinsci.plugins.workflow.actions.ErrorAction$ErrorId: 90dcd0ab-27e1-4cbc-b232-fe4d60ea8c94
01:02:32  hudson.remoting.ProxyException: java.nio.file.FileAlreadyExistsException: /var/lib/jenkins/jobs/scylla-master/jobs/scylla-ci/builds/17850/archive/allure-report.zip
01:02:32  	at java.base/sun.nio.fs.UnixFileSystem.move(UnixFileSystem.java:912)
01:02:32  	at java.base/sun.nio.fs.UnixFileSystemProvider.move(UnixFileSystemProvider.java:309)
01:02:32  	at java.base/java.nio.file.Files.move(Files.java:1431)
01:02:32  	at PluginClassLoader for allure-jenkins-plugin//ru.yandex.qatools.allure.jenkins.AllureReportPublisher.saveAllureArtifact(AllureReportPublisher.java:391)
01:02:32  	at PluginClassLoader for allure-jenkins-plugin//ru.yandex.qatools.allure.jenkins.AllureReportPublisher.generateReport(AllureReportPublisher.java:366)
01:02:32  	at PluginClassLoader for allure-jenkins-plugin//ru.yandex.qatools.allure.jenkins.AllureReportPublisher.perform(AllureReportPublisher.java:270)
01:02:32  	at PluginClassLoader for workflow-basic-steps//org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:101)
01:02:32  	at PluginClassLoader for workflow-basic-steps//org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:71)
01:02:32  	at PluginClassLoader for workflow-step-api//org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:49)
01:02:32  	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
01:02:32  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
01:02:32  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
01:02:32  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
01:02:32  	at java.base/java.lang.Thread.run(Thread.java:1583)
01:02:34  Finished: FAILURE
```

I'm not sure what is causing this.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17931/) | :white_check_mark: | i-0a3beb616cb47363f |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17931/) | :white_check_mark: | spider8.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17931/) | :white_check_mark: | spider8.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17931/)<br> :small_blue_diamond: test/boost/CMakeLists.txt<br> :small_blue_diamond: test/boost/bytes_ostream_test.cc<br> :small_blue_diamond: test/boost/comparable_bytes_test.cc | :white_check_mark: | spider8.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17931/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17931/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17931/)<br> :small_blue_diamond: test/boost/CMakeLists.txt<br> :small_blue_diamond: test/boost/bytes_ostream_test.cc<br> :small_blue_diamond: test/boost/comparable_bytes_test.cc | :white_check_mark: | spider1.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17931/) | :white_check_mark: | spider11.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17931/) | :x: | spider11.cloudius-systems.com |
| dtest | [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5727/) | :white_check_mark: |  |
|  | [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/8916/) | :white_check_mark: |  |
|  | [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/17008/) | :white_check_mark: |  |

#### Failed Tests (21/16308):
- [test_int.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17931/testReport/junit/boost.comparable_bytes_test/cc/debug___test_py___debug___test_int_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_int.debug.1)
- [test_timestamp.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17931/testReport/junit/boost.comparable_bytes_test/cc/debug___test_py___debug___test_timestamp_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_timestamp.debug.1)
- [test_varint.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17931/testReport/junit/boost.comparable_bytes_test/cc/debug___test_py___debug___test_varint_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_varint.debug.1)
- [test_varint_length_encoding.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17931/testReport/junit/boost.comparable_bytes_test/cc/debug___test_py___debug___test_varint_length_encoding_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_varint_length_encoding.debug.1)
- [test_decimal.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17931/testReport/junit/boost.comparable_bytes_test/cc/debug___test_py___debug___test_decimal_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_decimal.debug.1)
- [test_time.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17931/testReport/junit/boost.comparable_bytes_test/cc/debug___test_py___debug___test_time_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_time.debug.1)
- [test_bigint.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17931/testReport/junit/boost.comparable_bytes_test/cc/debug___test_py___debug___test_bigint_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_bigint.debug.1)
- [test_int.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17931/testReport/junit/boost.comparable_bytes_test/cc/debug___test_py___debug___test_int_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_int.debug.2)
- [test_timestamp.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17931/testReport/junit/boost.comparable_bytes_test/cc/debug___test_py___debug___test_timestamp_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_timestamp.debug.2)
- [test_time.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/17931/testReport/junit/boost.comparable_bytes_test/cc/debug___test_py___debug___test_time_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_time.debug.2)

#### Build Details:
- Duration: 3 hr 43 min

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17969/) | :white_check_mark: | thor |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17969/) | :white_check_mark: | i-0b5ea48eeb3019e1f |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17969/) | :white_check_mark: | i-0b5ea48eeb3019e1f |

#### Build Details:
- Duration: 36 min

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17977/) | :white_check_mark: | ran |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17977/) | :white_check_mark: | spider2.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/17977/) | :white_check_mark: | spider2.cloudius-systems.com |

#### Build Details:
- Duration: 1 hr 4 min

avikivity commented:
> Changelog V8:
> 
> * fixed issues in test that were causing failures in debug mode.

I don't see a debug mode CI report.

lkshminarayanan commented:
> > Changelog V8:
> > 
> > * fixed issues in test that were causing failures in debug mode.
> 
> I don't see a debug mode CI report.

Yeah. I'll retrigger the tests.

lkshminarayanan commented:
Rebased again to trigger the tests.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18056/) | :white_check_mark: | i-084d9cd4e2d8a41c5 |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18056/) | :white_check_mark: | spider4.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18056/) | :white_check_mark: | spider4.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18056/)<br> :small_blue_diamond: test/boost/CMakeLists.txt<br> :small_blue_diamond: test/boost/bytes_ostream_test.cc<br> :small_blue_diamond: test/boost/comparable_bytes_test.cc | :white_check_mark: | spider4.cloudius-systems.com |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18056/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18056/) | :white_check_mark: | spider1.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18056/)<br> :small_blue_diamond: test/boost/CMakeLists.txt<br> :small_blue_diamond: test/boost/bytes_ostream_test.cc<br> :small_blue_diamond: test/boost/comparable_bytes_test.cc | :white_check_mark: | spider1.cloudius-systems.com |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18056/) | :white_check_mark: | spider7.cloudius-systems.com |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18056/) | :white_check_mark: | spider7.cloudius-systems.com |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18056/)<br> :small_blue_diamond: test/boost/CMakeLists.txt<br> :small_blue_diamond: test/boost/bytes_ostream_test.cc<br> :small_blue_diamond: test/boost/comparable_bytes_test.cc | :white_check_mark: | spider7.cloudius-systems.com |
| dtest | [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/5825/) | :white_check_mark: |  |
|  | [dtest with consistent topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/9036/) | :white_check_mark: |  |
|  | [dtest with gossip topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/17127/) | :white_check_mark: |  |

#### Build Details:
- Duration: 5 hr 56 min
